Async workshop (#10)
tweedegolf/rust-workshop: slides/C-advanced-rust.md

C1 exercise description: [workshop.tweede.golf](https://workshop.tweede.golf/C1-parallel-rust/mod.html)

---
layout: cover
---

# Part 2
Async Rust

- Introduction Rust `async` programming
- Uncovering `Future` type
- Mechanics behind `async`/`await`
- Creating our own async primitives

---
layout: default
---

> # What is async rust?

Reviewer: ```suggestion
# What is async Rust?
```

Async function:

```rust
use tokio::fs;
use std::net::SocketAddr;

async fn read_address() -> Result<SocketAddr, Box<dyn std::error::Error + 'static>> {
    let contents = fs::read("address.txt").await?;
    let address: SocketAddr = String::from_utf8_lossy(&contents).parse()?;
    Ok(address)
}
```

This is sweet, but can we have less sugar?

```rust
use tokio::fs;
use std::net::SocketAddr;
use std::future::Future;

fn read_address() -> impl Future<Output = Result<SocketAddr, Box<dyn std::error::Error + 'static>>> {
    async {
        let contents = fs::read("address.txt").await?;
        let address: SocketAddr = String::from_utf8_lossy(&contents).parse()?;
        Ok(address)          
    }
}
```

---
layout: two-cols
---

# Async block?

Rust syntax

> Instructs the compiler to turn the code block into a statemachine

Reviewer: ```suggestion
Instructs the compiler to turn the code block into a state machine
```

```rust
let my_future = async {
    let a = rand();
    sleep().await;
    let b = a + rand();
    sleep().await;
    let c = (a + b + rand()).to_string();
    sleep().await;
    c
};
```

::right::

Only data carried over await points are kept.

> This statemachine would roughly look like:

Reviewer: ```suggestion
This state machine would roughly look like:
```

```rust
enum AnonymousFuture {
    Stage0 {
        a: u32
    },
    Stage1 {
        a: u32,
        b: u32,
    }
    Stage2 {
        c: String,
    }
}
```

> But how do we interact with the statemachine?

Reviewer: ```suggestion
But how do we interact with the state machine?
```

---
layout: default
---

# Future

Trait in the std

```rust
pub trait Future {
    type Output;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
}
```

> - Async statemachines implement `Future`

Reviewer: ```suggestion
- Async state machines implement `Future`
```

- A future moves forward by polling
- Keep polling until ready with output

```rust
pub enum Poll<T> {
    Ready(T),
    Pending,
}
```

- `Pin` guarantees `Self` isn't moved
- `poll` should not be called after it has previously returned `Ready`

---
layout: two-cols
---

# Polling a future

- The thing doing the polling is called the executor
- Continuously polling is inefficient

>   - Cpu needs to stay active

Reviewer: ```suggestion
  - CPU needs to stay active
```
